### PR TITLE
SC: fix a bug for delayed request nonce 0 handling case

### PR DESCRIPTION
**Pull request by jeonghonah**
## Proposed changes

-  A bug fix for handling 'request nonce > 0' before handling 'request nonce == 0' case.
- Add unit tests for checking `maxHandledRequestedNonce`, `sequentialHandleNonce` and `sequentialHandledRequestBlockNumber`
- The unit tests are based on the following document: 
https://docs.google.com/spreadsheets/d/1aQZVWTNGs6KLHKXUPxHDX5akGUBiGDS9xAexLG6zMfE/edit?pli=1#gid=184275023

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- #80, #139

### CODE REVIEW

File: contracts/bridge/BridgeTransferCommon.sol [hunk 1]
```diff
@@ -113,7 +113,12 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
         }
         for (i = sequentialHandleNonce; i <= maxHandledRequestedNonce && handledNoncesToBlockNums[i] > 0; i++) { }
         sequentialHandleNonce = i;
-        sequentialHandledRequestBlockNumber = handledNoncesToBlockNums[i-1];
+        // edge case: requestNonce zero is not yet handled.
+        if (sequentialHandleNonce == 0) {
+            sequentialHandledRequestBlockNumber = 1; // skip genesis block
+        } else {
+            sequentialHandledRequestBlockNumber = handledNoncesToBlockNums[i-1];
```
Comment (KimKyungup) [inline contracts/bridge/BridgeTransferCommon.sol]: ` sequentialHandledRequestBlockNumber = handledNoncesToBlockNums[i-1];` is important. So why don't we write only this? :)
If we want make initial sequentialHandledRequestBlockNumber to 1, we can set in declare code. :)

Comment (jeonghonah) [reply contracts/bridge/BridgeTransferCommon.sol]: Fixed.

File: node/sc/multi_bridge_test.go [hunk 2]
```diff
@@ -1006,3 +1006,117 @@ func TestMultiBridgeKLAYTransferMixConfig3(t *testing.T) {
 		}
 	}
 }
+
+// TestNoncesAndBlockNumber checks the following:
+// - set threshold to 2.
+// - the first transfer is done and check nonces and block number (req 0, blk 100000)
+// - the second transfer is done and check nonces and block number (req 1, blk 100100)
+func TestNoncesAndBlockNumber(t *testing.T) {
```
Comment (KimKyungup) [inline node/sc/multi_bridge_test.go]: Reverted sequence case can be added as a unit test. :)

Comment (jeonghonah) [reply node/sc/multi_bridge_test.go]: There is a TestNoncesAndBlockNumberUnordered case.

File: contracts/bridge/BridgeTransferCommon.sol [hunk 3]
```diff
@@ -11,10 +11,10 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
     bool public modeMintBurn = false;
     bool public isRunning;
 
-    uint64 public requestNonce;
-    uint64 public sequentialHandledRequestBlockNumber;
-    uint64 public sequentialHandleNonce;
-    uint64 public maxHandledRequestedNonce;
+    uint64 public requestNonce; // value transfer nonce increasing by 1 for each request.
+    uint64 public maxHandledRequestedNonce; // maximum handled requested nonce used for checking nonce update.
+    uint64 public sequentialHandleNonce; // the largest sequential handle nonce used for recovery.
```
[the remark this answers is not in the document]
Comment (KimKyungup) [reply contracts/bridge/BridgeTransferCommon.sol]: sequentialHandleNonce  -> minUnhandledRequestNonce?

File: contracts/bridge/BridgeTransferCommon.sol [hunk 4]
```diff
@@ -11,10 +11,10 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
     bool public modeMintBurn = false;
     bool public isRunning;
 
-    uint64 public requestNonce;
-    uint64 public sequentialHandledRequestBlockNumber;
-    uint64 public sequentialHandleNonce;
-    uint64 public maxHandledRequestedNonce;
+    uint64 public requestNonce; // value transfer nonce increasing by 1 for each request.
+    uint64 public maxHandledRequestedNonce; // maximum handled requested nonce used for checking nonce update.
```
Comment (KimKyungup) [inline contracts/bridge/BridgeTransferCommon.sol]: ```// maximum nonce of request event which is handled.```

File: contracts/bridge/BridgeTransferCommon.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeonghonah) [reply contracts/bridge/BridgeTransferCommon.sol]: Good. @aidan-kwon How about this?

File: contracts/bridge/BridgeTransferCommon.sol [hunk 5]
```diff
@@ -11,10 +11,10 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
...
+    uint64 public sequentialHandleNonce; // the largest sequential handle nonce used for recovery.
+    uint64 public sequentialHandledRequestBlockNumber = 1; // the largest sequential block number used for recovery.
```
Comment (jeonghonah) [inline contracts/bridge/BridgeTransferCommon.sol]: recoveryBlockNumber?

Comment (KimKyungup) [inline contracts/bridge/BridgeTransferCommon.sol]: `sequentialHandledRequestBlockNumber` -> `recoveryBlockNumber`
```// the block number which recovery start from.```

File: contracts/bridge/BridgeTransferCommon.sol [hunk 6]
```diff
@@ -11,10 +11,10 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
     bool public modeMintBurn = false;
     bool public isRunning;
 
-    uint64 public requestNonce;
-    uint64 public sequentialHandledRequestBlockNumber;
-    uint64 public sequentialHandleNonce;
-    uint64 public maxHandledRequestedNonce;
+    uint64 public requestNonce; // value transfer nonce increasing by 1 for each request.
```
[the remark this answers is not in the document]
Comment (aidan-kwon) [reply contracts/bridge/BridgeTransferCommon.sol]: // the number of value transfer request that the bridge contract received.

File: node/sc/bridge_test.go [hunk 7]
```diff
@@ -283,7 +283,7 @@ loop:
 				assert.NoError(t, err)
 				assert.Equal(t, bal, big.NewInt(int64(transferAmount*(testCount-nonceOffset+1))))
 
-				sequentialHandleNonce, err := b.SequentialHandleNonce(nil)
+				sequentialHandleNonce, err := b.MinUnhandledRequestNonce(nil)
```
Comment (KimKyungup) [inline node/sc/bridge_test.go]:
```suggestion
				minUnhandledRequestNonce, err := b.MinUnhandledRequestNonce(nil)
```

File: contracts/bridge/BridgeTransferCommon.sol [hunk 8]
```diff
@@ -12,10 +12,10 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
     bool public isRunning;
 
     uint64 public requestNonce; // the number of value transfer request that this contract received.
-    uint64 public minUnhandledRequestNonce; // a minimum nonce of a value transfer request that will be handled.
-    uint64 public maxHandledRequestedNonce; // maximum nonce of the counterpart bridge's value transfer request that is handled.
+    uint64 public lowerHandleNonce; // the nonce + 1 of a value transfer request surely processed by the counterpart bridge.
```
Comment (KimKyungup) [inline contracts/bridge/BridgeTransferCommon.sol]: How about keep the comment?
`// a minimum nonce of a value transfer request that will be handled.`

File: contracts/bridge/BridgeTransferCommon.sol [hunk 9]
```diff
@@ -12,10 +12,10 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
...
+    uint64 public upperHandleNonce; // the nonce of a value transfer request probably processed by the counterpart bridge.
```
Comment (KimKyungup) [inline contracts/bridge/BridgeTransferCommon.sol]: How about keep the comment?
```// maximum nonce of the counterpart bridge's value transfer request that is handled.```

**Pull request merged**